### PR TITLE
PLT-6214 Move channel store and actions over to redux

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Move channel store and actions over to redux. Waiting on #6222 but ready for PM review.

Testing areas:
* Creating/deleting/joining/leaving channels
* Channel user management (roles, removing, adding, etc.)
* Unreads between channels
* Updating channel notification settings
* Updating channel (header, purpose, etc.)
* Auto-linking channels
* Searching in a channel
* Anything else channel related you can think of

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6214

### CODE REVIEW

**Comment by jasonblais** [comment]
@jwilander the build seems to fail

**Comment by jwilander** [comment]
Yeah, I'm looking into it. Seems unrelated to my changes

**Comment by jasonblais** [comment]
Tested on Win10/Chrome. 

i'll have time later today to test channel notification preferences for email/mobile, gotta hop into something else now,

- Creating/deleting/joining/leaving channels
- Unreads (messages and mentions) across channels
- Updating channel (header, purpose, etc.)
- Auto-linking channels
- Searching in a channel
- Pinning posts, favoriting channels
- Max notifications settings
- Channel member count icon
- Channel user management (promoting/demoting from channel admin, removing, adding, etc.)
- Notification preferences for 'mark channels as unread' and 'desktop notifications'

1) Joining a channel post an incorrect system message:

Observed: `{username} added to the channel by {username}`

image

Expected: `{username} has joined the channel.`

2) Deleted channel appears bolded in the sidebar for other users
 - create a private channel
 - add other users to the channel
 - delete the channel

 observed: (for other users) Channel remains bolded in the sidebar. System message `jason has archived the channel.` is posted

image

 expected: (for other users) Channel disappears from the sidebar

3) Leaving a channel posts an incorrect system message
 - observed: `test was removed from the channel`
 - expected: `test has left the channel.`

4) Private channels appear in "more public channel" list giving a permission error when trying to join

image

image

5) Resetting channel URL doesn't close the "rename channel" modal properly
 - create a channel 
 - go to "Rename Channel" option
 - change the URL
 - hit "Save"
 - observed: no visible action; unable to close the modal without refresh (the URL does successfully change though)

6) Creating a DM fails and redirects you to Town Square
 - Click "More..." below DM list
 - Choose one member
 - Click "Go"
 - observed: User redirected to Town Square

7) Changing "Mark Channel Unread" only to mentions also changes "Send desktop notifications" to "Never" in the UI (although the desktop notification setting itself doesn't change)

image

EDIT: It seems the "Mark Channel Unread" option in the UI keeps changing based on "send desktop notification" setting, although the functionality doesn't actually change

**Comment by jasonblais** [comment]
Also found a bug that repros on master: https://mattermost.atlassian.net/browse/PLT-6414

**Comment by jwilander** [comment]
Thanks @jasonblais! I'll go through these and fix them a bit later

**Comment by jwilander** [comment]
@jasonblais fixed all above issues except for 7 which I wasn't able to reproduce

**Comment by jasonblais** [comment]
Confirm 1 - 6 are fixed.

Got better repro steps for 7

 - Set Channel Notification Preferences > Mark Channel Unread to `Only for mentions`
 - Save
 - Close the modal
 - reopen channel notification preferences

Observed: Send desktop notifications set to `Never` in the UI (although it's actually set to whatever it was before)

image

-----

Testing email/mobile push notifications right now,

**Comment by jasonblais** [comment]
8 - Setting one channel notification preference resets the other
 - set Send Desktop Notifications to 'only for mentions' (observed: notifications received for mentions only)
 - set Mark Channel Unread to 'only for mentions'
 - observed: Desktop notifications received for all messages. (The UI says `never` due to the bug described in 7)

Similarly, setting Send Desktop Notifications back to 'only for mentions' resets the Mark Channel Unread option. Clicking to edit the setting doesn't show either option being selected.

image

9 - Channel mobile push notification preferences don't work. They won't overwrite the account preferences.
 - have mobile push notification set 'for all activity' in Account Settings
 - change channel notification preferences to 'never'

Observed: Mobile push notifications for that channel keep coming. Same with different combinations of account/channel notification preferences.

(No issues with email notifications at the account-level - I realized we don't have channel-specific settings for them)

**Comment by jwilander** [comment]
Alright fixed 7,8,9. Thanks for the extra repro steps, that helped a lot

File: webapp/actions/channel_actions.jsx [hunk 1]
```diff
@@ -20,6 +20,29 @@ import {Constants, Preferences, ActionTypes} from 'utils/constants.jsx';
...
+    joinChannel as joinChannelRedux,
+    leaveChannel as leaveChannelRedux,
+    updateChannel as updateChannelRedux,
+    searchChannels,
+    updateChannelNotifyProps as updateChannelNotifyPropsRedux,
+    createChannel as createChannelRedux,
+    patchChannel,
+    getChannelMembersByIds,
+    deleteChannel as deleteChannelRedux
+} from 'mattermost-redux/actions/channels';
```
Comment (enahum) [inline webapp/actions/channel_actions.jsx]: Why not do something like `import * as ChannelActions from 'mattermost-redux/actions/channels'`? im really 0/5 but it seems simpler that way

File: webapp/actions/channel_actions.jsx [hunk 2]
```diff
@@ -283,11 +239,11 @@ export function unmarkFavorite(channelId) {
 }
 
 export function loadChannelsForCurrentUser() {
-    AsyncClient.getChannels().then(() => {
-        AsyncClient.getMyChannelMembers().then(() => {
+    fetchMyChannelsAndMembers(TeamStore.getCurrentId())(dispatch, getState).then(
```
Comment (enahum) [inline webapp/actions/channel_actions.jsx]: im also 0/5 on this but you could do

```js
fetchMyChannelsAndMembers(TeamStore.getCurrentId())(dispatch, getState).
then(loadDMsAndGMsForUnreads);
```

File: config/config.json [hunk 3]
```diff
@@ -70,7 +70,7 @@
     },
     "SqlSettings": {
         "DriverName": "mysql",
-        "DataSource": "mmuser:mostest@tcp(dockerhost:3306)/mattermost_test?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s",
```
Comment (crspeller) [inline config/config.json]: Unintended change.

**Comment by jwilander** [comment]
@crspeller fixed

**Pull request merged**